### ISSUE TITLE
Table columns resizing is not working correctly

### ISSUE THREAD

**Issue by agupta2-ibm-maximo** [carbon-design-system/carbon-addons-iot-react#1358]
Table columns resizing is not working correctly

Team,

We are observing an issue on table where column resizing is not working, As per our observations, Table is not able to fetch width of columns correctly and need to reload page to make it work sometimes even that did not work, Due to this column resizing feature is not working.
Please refer attached screenshots ("currentColumnWidths" object not able to populate width showing "undefined")
![WidthData](https://user-images.githubusercontent.com/56728062/86273793-64719f80-bbee-11ea-809c-12562bca3cba.JPG)
![NoWidthData](https://user-images.githubusercontent.com/56728062/86273797-663b6300-bbee-11ea-9981-f5de79dd9098.JPG)

**Comment by JordanWSmith15** [comment]
FYI, this environment shows it breaking: https://pages.github.ibm.com/maximo-app-framework/graphite/reference-app/#/tables

**Comment by davidicus** [comment]
There seems to be several props that are undefined. Do you know why these are not being set?

![image](https://user-images.githubusercontent.com/1590966/86277109-28b1e800-bba4-11ea-90de-a9c8f3bd3b2d.png)

**Comment by davidicus** [comment]
Thanks for reporting this bug. Please provide some more info on the options that are set for this Table.

**Comment by hsharma1-ibm-maximo** [comment]
Thanks for looking into the issue, Please refer attached screenshots of options we are setting, First screenshot (resize not working) and second screenshot (resize working). They both looks identical. Also we observe even if resizing work after page load clicking on one of sortable column also breaks resizing feature. Please suggest.

![notWorking](https://user-images.githubusercontent.com/67732023/86331406-afc79480-bc66-11ea-8203-0ea5a457d933.JPG)
![working](https://user-images.githubusercontent.com/67732023/86331414-b1915800-bc66-11ea-8460-aab92be1f6a3.JPG)

**Comment by hsharma1-ibm-maximo** [comment]
Steps to replicate - 
Storybook URL - https://pages.github.ibm.com/maximo-app-framework/graphite/react/storybook/

Scenario 1 -
1. Navigate to storybook using above URL.
2. Check "Can Resize" knob.
3. Perform resizing it should work.
4. Uncheck and re-check the "Can Resize" knob.
5. Perform resizing it did not work.

Scenario 2 -
1. Navigate to storybook.
2. Check "Can Resize" knob.
3. Perform resizing it should work.
4. Click any of the sortable column to sort the data.
5. Perform resizing it did not work.

**Comment by bjornalm** [comment]
I'm looking into this, but have a couple of questions.
1. Which version of the carbon-addons-iot-react library are you using?
2. Are you using the iot Table or the StatefulTable component?

**Comment by hsharma1-ibm-maximo** [comment]
Please refer below responses - 
1. carbon-addons-iot-react library version - 2.93.4
2. We are using table importing from library as - import {Table} from 'carbon-addons-iot-react';

**Comment by bjornalm** [comment]
Thanks. So you manage the state outside the table, how do you handle `onColumnResize`? Do you update your app state and use that state as input for the columns prop? I.e. something like

```
const [myColumns, setMyColumns] = useState(initialColumns);
const onColumnResize = cols => setMyColumns(cols);

return (
  <Table
    options={{
      hasResize: boolean('hasResize', true),
    }}
    columns={myColumns}
    data={tableData}
    actions={{ ...actions, table: { ...actions.table, onColumnResize } }}
    view={{
      table: {
        ordering: defaultOrdering,
      },
    }}
  />
);
```

**Comment by bjornalm** [comment]
I think I have figured out how this is happening. I managed to isolate and reproduce a similar behaviour locally. I'm not 100% this is the root cause of your issue but it seems likely given what I've seen in the attached screenshots. 

The resize fails when the table column prop is updated width a new array that has column objects with `width:undefined` after the table has been initially rendered. In my reproduced example I created a new column array containing objects with `width:undefined` and updated column the prop with that array when the sort was triggered, and the result was the same as described by this issue. 

So my hunch is that your state management or wrapping Datasource Table at some point does this, instead of setting the actual widths of the columns.

**Comment by stuckless** [comment]
While hopefully we can fix this by removing the width:undefined in our code... this obviously changed.   Before the upgrade the column resizing all worked (with the undefines), and now, it's basically a land mine of maybe it works or maybe it doesn't.  We'll try to delete the .width property for null/undefined widths and see what happens.

**Comment by bjornalm** [comment]
Sorry for the inconvenience. In order to dynamically load views containing modified column widths _after the initial rendering_ we had to make sure that the table handles changes to the column width properties. Before the upgrade any changes to the column width property after initial rendering was ignored, which made it possible for this to work even though the column prop was updated with `undefined` values for the visible columns.

**Comment by stuckless** [comment]
I get very scared, now, every time I see a commit against the Table, since I know something is going to break, but I can't ever predict what :(   I don't disagree with the change you had to make, just that the before/after behaviour is different, and as such, it's a breaking change.  So once a again semantic versioning isn't helping us determine that a build has breaking changes.   For the view manager... it really should have been that IF you have a view manager, you get this new behaviour, but, if you don't, then you get the preserved behaviour, since the feature was the addition of a view manager, which we are not using (at least not this one, since, we built our own).  Keeping the contract of semantic versioning is hard, and it usually means that when you add a feature that and you don't want to have a major version, you need to, in many cases have lots of branch code to preserve behaviours.  This is messy, but that's the contract of semantic versioning.

**Comment by bjornalm** [comment]
I totally understand. In hindsight this should have been a breaking change, but it can sometimes be very difficult to predict exactly how complex components like the table is being used. 

To clarify. It is still possible for the initial column prop to have objects with `width:undefined` and it is also possible for hidden columns to have `width:undefined` at any point but we did not consider that a resizable table would be getting a column prop with `width:undefined` for all the visible columns _after the initial rendering_. It is not an obvious scenario and I don't think it was an explicit choice to allow for that to work in the first place. Nevertheless, you are right about the contract of semantic versioning.

**Comment by hsharma1-ibm-maximo** [comment]
Hi Bjorn,
We observe an issue in adjusting column widths, here is the scenario - 
After initial table load,  Add some new columns using "Manage columns" feature.

Table resizing did not work on newly added columns.

Please refer below observation, let me know in case I am missing something or incorrect - 
In TableHead.jsx : 
useLayoutEffect function is responsible to calculate widths but it have an AND condition of "isEmpty(currentColumnWidths)" which works in initial load.
On using "Manage Columns" feature and adding new columns "isEmpty(currentColumnWidths)" returns "false" since it have older(initial load) data cached due to which width calculation for newly added columns (width:undefined) did not happen leading to inconsistent behavior of column resize feature.

Also, useDeepCompareEffect function have some code to update widths but it also bound by the condition "col.hasOwnProperty('width')" which in our case is false since we are expecting table component to update widths and are not passing any width attribute from our end.

Please suggest.

**Comment by bjornalm** [comment]
By "Manage Columns" and adding new columns, are you toggling a previously hidden column or are you actually adding a new column that previously has not been defined anywhere? I replicated a bug when toggling a previously hidden column that had no initial width. I have create a separate issue for that here #1419

The useDeepCompareEffect is used to allow the external modification of the columns, which is needed in the scenario where a user wants to load a previously saved view of his table that includes specific column widths.

**Comment by hsharma1-ibm-maximo** [comment]
I am adding a new column whose width was not previously calculated during initial rendering.

**Comment by bjornalm** [comment]
I'm trying to understand if you are _really adding_ a new column or just toggling a hidden one. By adding I mean that you after the initial rendering are actually programatically adding new items to the `column` prop and `view.table.ordering` prop (and possibly modifying the `data` prop) to reflect the new column.

By toggling hidden column I mean that you are using the show hide column function to simply show a column that has already been defined but was hidden on initial rendering (possibly without an explicit width, but still defined).

**Comment by hsharma1-ibm-maximo** [comment]
By adding I am referring new columns being passed using "column" props. Lets say initial column count is 5 (number of columns passed in "column" prop) and by using "Manage Columns" feature we are adding 2 more columns and passing total 7 columns in "column" prop.

**Comment by bjornalm** [comment]
@hsharma1-ibm-maximo Please have a look at the fix for your problem in this PR 
https://github.com/carbon-design-system/carbon-addons-iot-react/pull/1455

**Comment by hsharma1-ibm-maximo** [comment]
Hi @bjornalm 
Please refer below scenario-
Add a new hidden column ("newHiddenColumn") without providing width.
Add a new non-hidden column ("newColumn") without providing width
Remove hidden column ("newHiddenColumn").
Try resizing "newColumns"

Expected:
Resziing shoud happen.

Actual:
All columns started getting width as NaNpx and resizing did not work.

refer attached screenshot
![image](https://user-images.githubusercontent.com/67732023/89307992-5bfa0200-d68f-11ea-8466-83d13cd322d7.png)

**Comment by bjornalm** [comment]
@hsharma1-ibm-maximo Nice catch, I've updated the PR.

**Comment by davidicus** [comment]
I believe this bug is resolved from #1455  and #1420. If you still see it persisting please feel free to re-open
